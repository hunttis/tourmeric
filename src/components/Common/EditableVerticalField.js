import React, { Component } from 'react';
import firebase from 'firebase/app';
import { Translate } from 'react-localize-redux';
import _ from 'lodash';
import PropTypes from 'prop-types';

export default class EditableVerticalField extends Component {

  delayedSave = _.debounce((path, value) => {
    firebase.update(path, value); // Error handling?
    if (!this.unmounting) {
      this.setState({ saved: true, editing: false });
      this.delayedNormalize();
    }
  }, 1000)

  delayedNormalize = _.debounce(() => {
    if (!this.unmounting) {
      this.setState({ saved: false, editing: false });
    }
  }, 2000);

  constructor(props) {
    super(props);
    this.state = { saved: false, editing: false, fieldValue: props.defaultValue };
    this.unmounting = false;
  }

  componentDidMount() {
    this.unmounting = false;
  }

  componentWillUnmount() {
    this.unmounting = true;
  }

  handleChange(path, targetName, value) {
    this.setState({ editing: true, saved: false, fieldValue: value });
    this.delayedSave(path, { [targetName]: value });
  }

  render() {
    const {
      labelContent, placeHolder, defaultValue, path, targetName, inputType = 'text', idleIcon, emptyClass,
    } = this.props;
    const { saved, editing, fieldValue } = this.state;

    return (
      <div className="field">
        <label className="label">
          <Translate id={labelContent} />
        </label>
        <div className="field">
          <p className={`control is-expanded ${idleIcon ? 'has-icons-left' : 'has-icons-right'}`}>

            <Translate>
              {translate => (<input
                type={inputType}
                className={`input ${saved && 'is-success'} ${editing && 'is-warning'} ${(!editing && !saved) && 'is-normal'} ${emptyClass && !fieldValue ? 'is-danger' : ''}`}
                placeholder={translate(placeHolder)}
                defaultValue={defaultValue}
                onChange={event => this.handleChange(path, targetName, event.target.value)}
              />)
              }
            </Translate>
            {(idleIcon && !saved && !editing) && <span className="icon is-small is-left"><i className={`fas ${idleIcon}`} /></span>}
            {saved &&
              <span className={`icon is-small ${idleIcon ? 'is-left' : 'is-right'} ${emptyClass && !fieldValue ? 'has-text-danger' : 'has-text-success'}`}>
                <i className={`fas ${(emptyClass && !fieldValue) ? 'fa-thumbs-down' : 'fa-thumbs-up'}`} />
              </span>}
            {editing && <span className={`icon is-small ${idleIcon ? 'is-left' : 'is-right'} has-text-warning`}><i className="fas fa-pencil-alt" /></span>}
          </p>
        </div>
      </div>
    );
  }
}

EditableVerticalField.propTypes = {
  labelContent: PropTypes.string,
  placeHolder: PropTypes.string,
  defaultValue: PropTypes.string,
  path: PropTypes.string,
  targetName: PropTypes.string,
  inputType: PropTypes.string,
  idleIcon: PropTypes.string,
  emptyClass: PropTypes.string,
};
