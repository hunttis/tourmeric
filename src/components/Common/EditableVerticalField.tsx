import React, { Component } from 'react';
import firebase from 'firebase/app';
import { FormattedMessage } from "react-intl";
import _ from 'lodash';


interface Props {
  labelContent: string;
  placeHolder: string;
  defaultValue: string;
  path: string;
  targetName: string;
  inputType?: string;
  idleIcon?: string;
  emptyClass?: string;
  disabled?: boolean;
}

interface State {
  saved: boolean;
  editing: boolean;
  fieldValue: string;
}

export default class EditableVerticalField extends Component<Props, State> {

  delayedSave = _.debounce((path, value) => {
    firebase.update(path, value); // Error handling?
    if (!this.unmounting) {
      this.setState({ saved: true, editing: false });
      this.delayedNormalize();
    }
  }, 1000)

  delayedNormalize = _.debounce(() => {
    if (!this.unmounting) {
      this.setState({ saved: false, editing: false });
    }
  }, 2000);

  unmounting = false;

  constructor(props: Props) {
    super(props);
    this.state = { saved: false, editing: false, fieldValue: props.defaultValue };
    this.unmounting = false;
  }

  componentDidMount() {
    this.unmounting = false;
  }

  componentWillUnmount() {
    this.unmounting = true;
  }

  handleChange(path: string, targetName: string, value: string) {
    this.setState({ editing: true, saved: false, fieldValue: value });
    this.delayedSave(path, { [targetName]: value });
  }

  render() {
    const {
      labelContent, placeHolder, defaultValue, path, targetName, inputType = 'text', idleIcon, emptyClass, disabled = false,
    } = this.props;
    const { saved, editing, fieldValue } = this.state;

    return (
      <div className="field">
        {labelContent &&
          <label className={`label ${disabled && 'has-text-info'}`}>
            <FormattedMessage id={labelContent} />
          </label>
        }
        <div className="field">
          <p className={`control is-expanded ${idleIcon ? 'has-icons-left' : 'has-icons-right'}`}>

            
              <input
                type={inputType || 'text'}
                className={`input ${saved && 'is-success'} ${editing && 'is-warning'} ${(!editing && !saved) && 'is-normal'} ${emptyClass && !fieldValue ? 'is-danger' : ''} ${disabled && 'has-text-info'}`}
                placeholder={`${translate(placeHolder)}`}
                defaultValue={defaultValue}
                onChange={(event) => this.handleChange(path, targetName, event.target.value)}
                disabled={disabled}
              />
            
            {(idleIcon && !saved && !editing) && <span className="icon is-small is-left"><i className={`fas ${idleIcon} ${disabled && 'has-text-black'}`} /></span>}
            {saved &&
              <span className={`icon is-small ${idleIcon ? 'is-left' : 'is-right'} ${emptyClass && !fieldValue ? 'has-text-danger' : 'has-text-success'}`}>
                <i className={`fas ${(emptyClass && !fieldValue) ? 'fa-thumbs-down' : 'fa-thumbs-up'}`} />
              </span>}
            {editing && <span className={`icon is-small ${idleIcon ? 'is-left' : 'is-right'} has-text-warning`}><i className="fas fa-pencil-alt" /></span>}
          </p>
        </div>
      </div>
    );
  }
}
