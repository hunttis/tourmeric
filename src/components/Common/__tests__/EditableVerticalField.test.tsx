import React from 'react';
import { shallow } from 'enzyme';
import EditableVerticalField from '../EditableVerticalField';

describe('EditableVerticalField Tests', () => {

  it('Compares EditableVerticalField to snapshot', () => {

    const editableField = shallow(<EditableVerticalField
      labelContent="label"
      placeHolder="placeholder"
      defaultValue="defaultvalue"
      path="path"
      targetName="target"
      inputType="input"
      idleIcon=""
      emptyClass=""
      disabled={false}
    />);

    expect(editableField.exists()).toBe(true);
    expect(editableField).toMatchSnapshot();
  });

  // TODO: Write a test for the save and normalize functions
  // TODO: If possible, write tests for the mount and unmounting as well

});
