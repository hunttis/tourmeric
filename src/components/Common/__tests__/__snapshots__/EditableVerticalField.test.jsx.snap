// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableVerticalField Tests Compares EditableVerticalField to snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditableVerticalField
    defaultValue="defaultvalue"
    inputClasses="inputClasses"
    inputType="input"
    labelContent="label"
    leftIcon="lefticon"
    path="path"
    placeHolder="placeholder"
    targetName="target"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          className="label"
        >
          <Translate
            id="label"
          />
        </label>,
        <div
          className="field"
        >
          <p
            className="control is-expanded has-icons-right"
          >
            <Translate>
              [Function]
            </Translate>
          </p>
        </div>,
      ],
      "className": "field",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Translate
            id="label"
          />,
          "className": "label",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "id": "label",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <p
            className="control is-expanded has-icons-right"
          >
            <Translate>
              [Function]
            </Translate>
          </p>,
          "className": "field",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Translate>
                [Function]
              </Translate>,
              undefined,
              false,
              false,
            ],
            "className": "control is-expanded has-icons-right",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": [Function],
              },
              "ref": null,
              "rendered": [Function],
              "type": [Function],
            },
            undefined,
            false,
            false,
          ],
          "type": "p",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            className="label"
          >
            <Translate
              id="label"
            />
          </label>,
          <div
            className="field"
          >
            <p
              className="control is-expanded has-icons-right"
            >
              <Translate>
                [Function]
              </Translate>
            </p>
          </div>,
        ],
        "className": "field",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Translate
              id="label"
            />,
            "className": "label",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "id": "label",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p
              className="control is-expanded has-icons-right"
            >
              <Translate>
                [Function]
              </Translate>
            </p>,
            "className": "field",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Translate>
                  [Function]
                </Translate>,
                undefined,
                false,
                false,
              ],
              "className": "control is-expanded has-icons-right",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": [Function],
                },
                "ref": null,
                "rendered": [Function],
                "type": [Function],
              },
              undefined,
              false,
              false,
            ],
            "type": "p",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
